Change fromFuture to take a cats.Eval param

alexandru opened the pull request:
~~This is a proposal to rename `fromFuture` to `deferFuture`.~~

Scala is by default an eager language, which is why one of the first WTFs of people learning Scalaz is in noticing that by-name parameters are used everywhere. And then the second WTF you have is when moving from Scalaz to Cats and noticing that everything is eager by default, as it should be, since this is Scala.

The most obvious example is `Applicative.pure`, which in Cats takes an eager parameter. Now this is problematic because the lazy by-name variant doesn't exist yet, but given that this is Scala, having an eager `pure` makes a lot of sense and avoids a lot of gotchas.

By-name parameters are problematic in themselves, because changing the code a little, you get a different behavior. This has been the rationale for pushing `Eval` in Cats and recommending it for usage instead of by-name params and `lazy val`s.

~~So, the reason for why I would prefer a `deferFuture` is because the name makes it more visible that this function gets a by-name parameter, instead of a strict value.~~

~~In my mind a `fromFuture` function, in the context of cats, should take a strict value. Now you may not want an eager `fromFuture` in `IO`, which is totally fine, but I would still prefer `deferFuture` (or `suspendFuture`?) for the explicitness (I know, I know, some folks prefer to only read the types and not care about descriptive names, I'm not one of those people :-P).~~

This PR makes the param of `fromFuture` a `cats.Eval`, thus allowing the user to control the evaluation (e.g. eager, lazy, memoized).

Full list of changes:

- changed signature: `def fromFuture[A](f: cats.Eval[Future[A]]): IO[A]`
- added tests, refactored `IOTests` a little
- made the `Eq[IO]` depend on `TestContext`, which now makes it possible to test for `IO.never` ;-)
- moved `Eq[IO]` and `Eq[Future]` test instances into `cats.effect.laws.util.TestInstances` for reuse

I don't mind if you don't like it, I'm not dying on this hill :-)

pchlupacek commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -533,13 +533,13 @@ object IO extends IOInstances {
    * @see [[IO#unsafeToFuture]]
    */
-  def fromFuture[A](f: => Future[A])(implicit ec: ExecutionContext): IO[A] = {
+  def deferFuture[A](f: => Future[A])(implicit ec: ExecutionContext): IO[A] = {
```
I would rather instead define this as fromFuture, but instead of lazy argument I would propose the future to be wrapped in `F`, like : `fromFuture(f: F[Future[A]]`. Users then can use this as either `fromFuture(IO.now(Future)` or `fromFuture(IO.delay(future))`, where from code is then very clearly seen whether the future is delayed or not.

alexandru replied:
If we go in this direction, might as well use `cats.Eval`, since it would make better sense and was designed precisely for this purpose.

alexandru replied:
E.g.

```scala
import cats.Eval.always

IO.fromFuture(always(Future(???)))
```

pchlupacek replied:
Generally no problem with this. Although my first feeling is that Eval is too powerful for this use case (not sure if you will ever need Eval.later or Eval.deferred). Was hoping implementation will be that simple as f.flatMap { future => /** current code **/ }

alexandru replied:
Controlling evaluation (eager / lazy / memoized) is why `Eval` was introduced in Cats. That it evolved into an equivalent of `scalaz.effect.IO`, well, that happens :-)

In general you can't rely on `flatMap` to suspend side effects, as the law, even though in practice it happens, as for example there are implementations that optimize `F.pure(())` by making it eager.

Also, the point of renaming this to `deferFuture` was that some use-cases will need eager evaluation, fewer that those needing deferred evaluation that's true, but Scala is an eager language in which people do dirty things, so it will happen.

And so my general point here is that I'd like this to be named `deferFuture` to make it clear that the parameter is lazy, because otherwise some people will assume it is eager and that can also be dangerous.

mpilquist commented:
I have no strong opinions on this one -- `deferFuture` is fine with me. I like the gentle reminder that folks probably should include future construction inside the by-name. We could also have a `evalFuture` which takes an `Eval[Future[A]]` though I could live without that too.

pchlupacek commented:
I can sort of live with whatever. Just to me is bad habit having in API half of the arguments eager and another half be strict. Thats why I will be happy with strict F[_]. 

To me Eval is one more typeclass to understand for user, and thanks to its overlapping nature with IO I wouldn't force users to use it. IO would just work fine here, and I cannot imagine (maybe having lack of imagination) where IO won't be sufficient as `F`. 

Don't have any strong opinions here, was just merely suggestion. I think hence we have `IO.eval` already, people that really like `Eval` may easily get it (like IO.fromFuture(IO.eval(Future(...))). 

If any other typeclass, I would suggest to have `Sync` there.

alexandru commented:
This `IO` is async and too strong to be used as a type for controlling evaluation, plus it doesn't do memoization yet ... which would be a useful operation, but given my experience of implementing it in Monix's `Task`, it gets tricky and I'm not sure we want that kind of concurrency in `IO`. 

Whereas `Eval` just does `later` and `memoize`, because they are basically for free.

Of course, this is a useful discussion to have ... does `Eval` make sense in the presence of this `IO`? I think it does, for instance `Eval` is a `Comonad`, whereas `IO` shouldn't be (i.e. admitting an `unsafeRunSync` in its implementation is enough of a compromise, but we can't use that willy-nilly). And if we have `Eval`, might as well use it for its intended purpose.

But then boxing stuff in both `Eval` and `IO` is maybe a little overkill. Hence I would be happy with my `deferFuture` :-)

djspiewak commented:
General thoughts…

I agree that having random by-name parameter is weird, and also atypical for a cats-derived project.  The proper idiom would be to use `Eval` instead, as has been mentioned.  I think this is a really good idea.

I'm not fond of `deferFuture`, since it doesn't reflect the transformation *to* an `IO`.  `fromFuture` really seems like a better name to me, though it's not a hill I will die on.  I think that if the type signature was `Eval[Future[A]] => IO[A]` instead of `(=> Future[A]) => IO[A]`, the `defer` argument becomes less compelling.

I'm not opposed to parameterizing the effect here, but unfortunately we don't have any way of doing this *usefully* without using `Effect`, which as @alexandru pointed out, would be massive overkill for the goal of simply providing deferred execution.

I think the correct solution would be to change the by-name to an `Eval`.  I would be in favor of leaving the name as-is, but if everyone else prefers a different name, I will demur.

djspiewak commented on laws/shared/src/test/scala/cats/effect/BaseTestsSuite.scala:
```diff
@@ -17,15 +17,82 @@
+   * Defines equality for `IO` references that can
+   * get interpreted by means of a [[TestContext]].
+   */
+  implicit def eqIO[A](implicit A: Eq[A], ec: TestContext): Eq[IO[A]] =
```
I would be in favor of moving this to `src/main` in **cats-effect-laws**, though obviously not marked as `implicit`.  The reason being that almost anyone who wants to run the laws is going to need to define an `Eq[IO[A]`, and this is rapidly expanding into something which cannot be trivially replicated.

alexandru replied:
OK, but you noticed the signature, right? 

We depend on a `TestContext` for simulating async evaluation.

djspiewak replied:
Yes, I do see that.  Obviously `TestContext` would have to move out as well as supporting infrastructure.  I'm thinking something like `cats.effect.laws.util` or something like that.

I just want to make sure it's as frictionless as possible to use these laws on downstream projects.  I did some playing around with applying the laws to other types over the weekend and noticed that one of the most annoying steps was having to constantly redefine `Eq[IO[A]]`.

alexandru commented:
If everybody is in favour, we can do the `Eval` signature.

Also, we already have `eval`, so can we name it `evalFuture` 😆

djspiewak commented:
> Also, we already have eval, so can we name it evalFuture 😆

🤷‍♂️  I'm ok with that.  It's better than `deferFuture`, IMO.  I still like `fromFuture` better though.  cats method names generally don't explicitly reference the fact that they use `Eval` (e.g. `Foldable`'s `foldRight`); they just… use it.

pchlupacek commented:
+1 for evalFuture // fromFuture or simple `IO.future(f:Eval[Future[A]])` :-)

alexandru commented:
> cats method names generally don't explicitly reference the fact that they use Eval

That's not a general rule, there's also `catchNonFatal` versus `catchNonFatalEval` in `ApplicativeError`, so sometimes "eval" happens for visibility.

If you want `fromFuture` with an `Eval`, that addresses my concerns. I would prefer `evalFuture`, but you guys take a vote :)

@mpilquist what say you?

mpilquist commented:
👍 to `Eval` based signature. Name-wise, I slightly prefer `fromFuture` over `evalFuture` mostly to avoid confusion with `eval`.

alexandru commented:
Changes:

1. renamed `deferFuture` to `fromFuture`
2. made its param a `cats.Eval`
3. added test for protecting against `always(throw ex)`
4. moved `Eq[IO]` and `Eq[Future]` test instances into `cats.effect.laws.util.TestInstances`

djspiewak commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -533,19 +550,21 @@ object IO extends IOInstances {
    * {{{
-   * IO.fromFuture(f).unsafeToFuture === f     // true-ish (except for memoization)
-   * IO.fromFuture(ioa.unsafeToFuture) === ioa // true!
+   *   IO.fromFuture(now(f)).unsafeToFuture === f
```
Still needs the memoization caveat.

djspiewak commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -533,19 +550,21 @@ object IO extends IOInstances {
 
-      f onComplete {
-        case Failure(t) => cb(Left(t))
+      try f.value onComplete {
```
This `try`/`catch` is unnecessary.

alexandru commented:
@djspiewak I have addressed your comments.
